net/devif/devif.h

```diff
@@ -260,6 +260,9 @@ typedef CODE uint16_t (*devif_callback_event_t)(FAR struct net_driver_s *dev,
 struct devif_callback_s
 {
   FAR struct devif_callback_s *nxtconn;
+  FAR struct devif_callback_s *tail_helper; /* The tail helper field is used
```

Comment:
Hi Xiang,
Initially (for simplicity) just to check if FIFO vs LIFO helps to resolve the root issue, I applied a loop to find the tail in the list (I did not commit that code). As I made sure it resolved the issue, I improved the code and eliminated the loop by adding the tail helper. Thus you can see my current PR is w/o the loop.
The list is normally short only during low network traffic. During high network traffic the list grows towards to the maximum length, thus it is normally long instead.
Considering the process of adding items to the list, the time complexity was O(n) in LIFO mode.
If the loop is applied to find the list tail each time when a new item is added to the list, then the time complexity becomes near to O(n^2).
As I avoided applying the loop, the time complexity still stays O(n) as it was before my change.
Thus I think it's not good to make it O(n^2) to degrade the total performance (it's especially important in case of high network traffic).

Reply:
Ok, but is net_driver_s more suitable to add the tail pointer?

Reply:
Actually, before introducing the tail helper I was thinking about applying sq_queue_s as an ideal approach for this PR, including unification reasons. However, it seems it's not easy to do because devif_callback_s structure has links to two different nested chains (lists):
```
struct devif_callback_s
{
  FAR struct devif_callback_s *nxtconn;
  FAR struct devif_callback_s *nxtdev;
...
```
And it would be required to rewrite many places in the whole network protocol stack to adapt the code to the sq_queue_s.

Then I was trying to introduce a new structure describing only d_conncb list of net_driver_s structure as follows:
```
struct devif_callback_list_s
{
  FAR struct devif_callback_s *head;
  FAR struct devif_callback_s *tail;
};
```
However, in this case a huge effort is still required to adapt many parts of the whole network protocol stack, including net/bluetooth, net/can, net/ieee802154 and net/sixlowpan that I can not test by myself (I do not an appropriate hardware setup for that). Thus I decided to stop that effort. And I ended up with introducing tail_helper field to devif_callback_s structure as a workaround (the current PR). That helped to avoid adapting the whole network protocol stack. Actually I've modified only two files: net/devif/devif.h and net/devif/devif_callback.c.

Considering your idea, as I understand you mean to add something like this, correct? :
```
diff --git a/include/nuttx/net/netdev.h b/include/nuttx/net/netdev.h
index fe61124e7c..f9936364c3 100644
--- a/include/nuttx/net/netdev.h
+++ b/include/nuttx/net/netdev.h
@@ -383,6 +383,7 @@ struct net_driver_s
    */
 
   FAR struct devif_callback_s *d_conncb;
+  FAR struct devif_callback_s *d_conncb_tail;
   FAR struct devif_callback_s *d_devcb;
```

Reply:
Yes, to insert devif_callback_s by order, it's enough to record the list tail.

Reply:
In this case I will have to add an extra argument (devif_callback_s *d_conncb_tail) into devif_callback_alloc() and devif_conn_callback_free() functions and pass the tail pointer in all places where the functions are invoked.
If it's acceptable, I can do this. This would be still a workaround, however the advantage is it's more effective concerning memory usage. The effort is expected to be not so high compared to applying sq_queue_s or devif_callback_list_s structure. And I think it should not potentially break some functionality like net/bluetooth, net/can, net/ieee802154, net/sixlowpan.
Thus if this way is going to be accepted, please let me know.

Reply:
why need pass d_conncb_tail? since devif_callback_alloc/devif_callback_free already accept net_driver_s as the first argument, we can get the d_conncb_tail from the first argument.

Reply:
> why need pass d_conncb_tail? since devif_callback_alloc/devif_callback_free already accept net_driver_s as the first argument, we can get the d_conncb_tail from the first argument.

Though devif_callback_alloc/devif_callback_free accept net_driver_s as the first argument, this first argument (*dev) is passed as NULL in multiple places where the functions are invoked, while the other argument (**list) is effective (not NULL).

Reply:
In this case, how about search to the end?

Reply:
Possibly it's reasonable to add an option into menuconfig to switch the optimization way: memory/performance? And add conditonal compilation depending on the option?

Reply:
If we just add a tail pointer in net_driver_s, the size increment is very little(~10B) since the instance of net_driver_s is normally 1~2 on most IoT devices. So, it's more simpler to always enable this feature.

Reply:
Possibly you mean adding a tail pointer into devif_callback_s structure? (as I've already done in the PR)
If a tail pointer is added into net_driver_s structure, we have to add an extra argument into devif_callback_alloc/devif_callback_free functions, as the first argument (*dev) can not be used to get the tail pointer.

Reply:
> Possibly you mean adding a tail pointer into devif_callback_s structure? (as I've already done in the PR)

No, I mean adding the tail pointer to net_driver_s only. If we only want to add an entry to the head or tail of list, the single link list is enough to achieve O(1) performance.

> If a tail pointer is added into net_driver_s structure, we have to add an extra argument into devif_callback_alloc/devif_callback_free functions, as the first argument (*dev) can not be used to get the tail pointer.

Yes, we have to fallback to the linear search(O(n)), but it could be acceptable if NULL pointer happen rarely.

The current change look very strange, you add a pointer field to the list entry which basically become a double link list, but the code is still used it as a single link list.

Reply:
> Yes, we have to fallback to the linear search(O(n)), but it could be acceptable if NULL pointer happen rarely.

At least usrsock_setup_request_callback() and usrsock_pollsetup() always pass NULL pointer, that is not rarely.

Thus do you mean to branch depending on the first argument (*dev) value as follows?
1) If *dev == NULL then do linear search for the list tail
2) else get the tail pointer from net_driver_s structure

Reply:
Yes, or we switch the full double link list.

Reply:
In case of doubly linked list there is still a need to search for the list tail to implement FIFO instead of LIFO. The tail search will be needed in devif_conn_event() instead of devif_callback_alloc().

Reply:
The full double link list have all info:
```
struct net_driver_s
{
  FAR devif_callback_s *head;
  FAR devif_callback_s *tail;
};

struct devif_callback_s
{
  FAR devif_callback_s *prev;
  FAR devif_callback_s *next;
}
```

Reply:
> In case of doubly linked list there is still a need to search for the list tail to implement FIFO instead of LIFO. 

Why need? the doubly linked list should get the tail pointer directly.

`> The tail search will be needed in devif_conn_event() instead of devif_callback_alloc(). Thus doubly linked list is worse compared to my current PR concerning the performance, while memory usage would be the same as in my PR.

BTW, there still has an O(n) loop in devif_callback_free need to optimize:
https://github.com/apache/incubator-nuttx/pull/4433/files#diff-8eafb3edcad89f54bf9d2424eb4819a571b5d839a55d3eef8e679cfc76e5e820L94-L98
https://github.com/apache/incubator-nuttx/pull/4433/files#diff-8eafb3edcad89f54bf9d2424eb4819a571b5d839a55d3eef8e679cfc76e5e820R123-R127

Reply:
> Why need? the doubly linked list should get the tail pointer directly.

By "doubly linked list" do you mean adding *prevconn pointer like this?
```
diff --git a/net/devif/devif.h b/net/devif/devif.h
index deb7d9c29a..4dfb168ae2 100644
--- a/net/devif/devif.h
+++ b/net/devif/devif.h
@@ -260,6 +260,7 @@ typedef CODE uint16_t (*devif_callback_event_t)(FAR struct net_driver_s *dev,
 struct devif_callback_s
 {
   FAR struct devif_callback_s *nxtconn;
+  FAR struct devif_callback_s *prevconn;
   FAR struct devif_callback_s *nxtdev;
   FAR devif_callback_event_t event;
   FAR void *priv;
```
Or you mean applying e.g. dq_queue_s + dq_entry_s ?

Reply:
> BTW, there still has an O(n) loop in devif_callback_free need to optimize:
> https://github.com/apache/incubator-nuttx/pull/4433/files#diff-8eafb3edcad89f54bf9d2424eb4819a571b5d839a55d3eef8e679cfc76e5e820L94-L98
> https://github.com/apache/incubator-nuttx/pull/4433/files#diff-8eafb3edcad89f54bf9d2424eb4819a571b5d839a55d3eef8e679cfc76e5e820R123-R127

That is a good point. If we turn the list into a doubly linked list, the loop in devif_callback_free() can be eliminated.

Reply:
> > Why need? the doubly linked list should get the tail pointer directly.
> 
> By "doubly linked list" do you mean adding *prevconn pointer like this?
> 
> ```
> diff --git a/net/devif/devif.h b/net/devif/devif.h
> index deb7d9c29a..4dfb168ae2 100644
> --- a/net/devif/devif.h
> +++ b/net/devif/devif.h
> @@ -260,6 +260,7 @@ typedef CODE uint16_t (*devif_callback_event_t)(FAR struct net_driver_s *dev,
>  struct devif_callback_s
>  {
>    FAR struct devif_callback_s *nxtconn;
> +  FAR struct devif_callback_s *prevconn;
>    FAR struct devif_callback_s *nxtdev;
>    FAR devif_callback_event_t event;
>    FAR void *priv;
> ```
> 
> Or you mean applying e.g. dq_queue_s + dq_entry_s ?

Either is fine. Basically, we have two different approach:

1. Keep the single link list as before, but add the tail pointer to struct devif_callback_s or create a new devif_callback_list_s with head and tail pointer, then add/remove the node from the head/tail is O(1) operation, but free still O(n).
2. Change two the double link list, all operation can be finished in O(1).

Reply:
Let me explain possible approaches from my current point of view.

**Before the PR the situation is as follows:**
- The queue (list) is LIFO, however it should be FIFO.
- The list is singly linked.
- The list has a head pointer (inside of outer net_driver_s structure), however it does not have a tail pointer.
- The list item is devif_callback_s structure. It has two pointers to two different list chains (*nxtconn and *nxtdev) that seriously complicates applying dq_queue_s + dq_entry_s.
- The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL).  
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
- devif_conn_event() time complexity is O(n).

**Approach 1.s (the current state of the PR):**
- The queue (list) is FIFO as it should be.
- The list is singly linked.
- The list has a head pointer (inside of outer net_driver_s structure), and it has a tail pointer field inside of each list item as a workaround.
- The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL).  
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- Only two files are modified (devif.h and devif_callback.c). No need to adapt the whole network protocol stack.

**Approach 2.s:**
- The queue (list) is FIFO as it should be.
- The list is singly linked.
- The list has a head pointer (inside of outer net_driver_s structure), and a tail pointer is to be added into outer net_driver_s structure.
- The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL). If (*dev) argument is not NULL, the list tail can be taken from outer net_driver_s structure through (*dev) argument. Otherwise, the list tail can be found using a linear search (at least usrsock_setup_request_callback() and usrsock_pollsetup() always pass NULL pointer into (*dev) argument).
- devif_callback_alloc() time complexity is O(1) or O(n) in the worst case (i.e. O(n) and O(n^2) to fill the whole list).
- devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- Only two files are to be modified (netdev.h and devif_callback.c). No need to adapt the whole network protocol stack.

**Approach 3.s:**
- The queue (list) is FIFO as it should be.
- The list is singly linked.
- The list has a head pointer (inside of outer net_driver_s structure), and a tail pointer is to be added into outer net_driver_s structure.
- The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL).
- An extra (*tail) argument is to be added to devif_callback_alloc() function.
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- Multiple files are to be modified. Though a risk to break something is low.

**Approach 4.s:**
- The queue (list) is FIFO as it should be.
- The list is singly linked.
- A new devif_callback_list_s structure holding both the list head and tail is to be added into outer net_driver_s structure.
- The list item is still devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- (*list) argument type is to be changed to the new devif_callback_list_s in all list functions.
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- A lot of files are to be modified. Moreover, a risk to break something is high.

**Approach 5.s:**
- The queue (list) is FIFO as it should be.
- The list is singly linked by applying dq_queue_s + dq_entry_s.
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- A lot of files are to be modified. Moreover, a risk to break something is high. And so far It's not clear how to handle two different list chains (*nxtconn and *nxtdev) in this approach.

**Approach 1.d:**
- The queue (list) is FIFO as it should be.
- The list is doubly linked (an extra (*prevconn) field is to be added into devif_callback_s structure).
- The list has a head pointer (inside of outer net_driver_s structure), and it has a tail pointer field inside of each list item as a workaround.
- The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL).  
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- Only two files are modified (devif.h and devif_callback.c). No need to adapt the whole network protocol stack.

**Approach 2.d:**
- The queue (list) is FIFO as it should be.
- The list is doubly linked (an extra (*prevconn) field is to be added into devif_callback_s structure).
- The list has a head pointer (inside of outer net_driver_s structure), and a tail pointer is to be added into outer net_driver_s structure.
- The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL). If (*dev) argument is not NULL, the list tail can be taken from outer net_driver_s structure through (*dev) argument. Otherwise, the list tail can be found using a linear search (at least usrsock_setup_request_callback() and usrsock_pollsetup() always pass NULL pointer into (*dev) argument).
- devif_callback_alloc() time complexity is O(1) or O(n) in the worst case (i.e. O(n) and O(n^2) to fill the whole list).
- devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- Only three files are to be modified (netdev.h, devif.h and devif_callback.c). No need to adapt the whole network protocol stack.

**Approach 3.d:**
- The queue (list) is FIFO as it should be.
- The list is doubly linked (an extra (*prevconn) field is to be added into devif_callback_s structure).
- The list has a head pointer (inside of outer net_driver_s structure), and a tail pointer is to be added into outer net_driver_s structure.
- The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL).
- An extra (*tail) argument is to be added to devif_callback_alloc() function.
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- Multiple files are to be modified. Though a risk to break something is low.

**Approach 4.d:**
- The queue (list) is FIFO as it should be.
- The list is doubly linked (an extra (*prevconn) field is to be added into devif_callback_s structure).
- A new devif_callback_list_s structure holding both the list head and tail is to be added into outer net_driver_s structure.
- The list item is still devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
- (*list) argument type is to be changed to the new devif_callback_list_s in all list functions.
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- A lot of files are to be modified. Moreover, a risk to break something is high.

**Approach 5.d:**
- The queue (list) is FIFO as it should be.
- The list is doubly linked by applying dq_queue_s + dq_entry_s.
- devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
- devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
- devif_conn_event() time complexity is O(n).
- A lot of files are to be modified. Moreover, a risk to break something is high. And so far It's not clear how to handle two different list chains (*nxtconn and *nxtdev) in this approach.

Do you agree with all details of the outlined explanation?

Reply:
> Let me explain possible approaches from my current point of view.
> 

@a-lunev thanks for the detailed explanation.

> **Approach 1.s (the current state of the PR):**
> 

> 
> **Approach 2.s:**
> 
> * The queue (list) is FIFO as it should be.
> * The list is singly linked.
> * The list has a head pointer (inside of outer net_driver_s structure), and a tail pointer is to be added into outer net_driver_s structure.
> * The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
> * The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL). If (*dev) argument is not NULL, the list tail can be taken from outer net_driver_s structure through (*dev) argument. Otherwise, the list tail can be found using a linear search (at least usrsock_setup_request_callback() and usrsock_pollsetup() always pass NULL pointer into (*dev) argument).
> * devif_callback_alloc() time complexity is O(1) or O(n) in the worst case (i.e. O(n) and O(n^2) to fill the whole list).
> * devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
> * devif_conn_event() time complexity is O(n).
> * Only two files are to be modified (netdev.h and devif_callback.c). No need to adapt the whole network protocol stack.
> 

The problem of Approach 1.s is that it consume the same memory as the doubly linked list, but the O(n) complexity still happen in devif_callback_free.

> 
> **Approach 4.s:**
> 
> * The queue (list) is FIFO as it should be.
> * The list is singly linked.
> * A new devif_callback_list_s structure holding both the list head and tail is to be added into outer net_driver_s structure.
> * The list item is still devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
> * (*list) argument type is to be changed to the new devif_callback_list_s in all list functions.
> * devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
> * devif_callback_free() time complexity is O(n) (i.e. O(n^2) to empty the whole list).
> * devif_conn_event() time complexity is O(n).
> * A lot of files are to be modified. Moreover, a risk to break something is high.
> 

It look like the best approach with the singly linked list.

> **Approach 1.d:**
> 
> * The queue (list) is FIFO as it should be.
> * The list is doubly linked (an extra (*prevconn) field is to be added into devif_callback_s structure).
> * The list has a head pointer (inside of outer net_driver_s structure), and it has a tail pointer field inside of each list item as a workaround.
> * The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
> * The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL).
> * devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
> * devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
> * devif_conn_event() time complexity is O(n).
> * Only two files are modified (devif.h and devif_callback.c). No need to adapt the whole network protocol stack.
> 

The additional memory consumption is 3x, not 2x, not good as expect.

> **Approach 2.d:**
> 
> * The queue (list) is FIFO as it should be.
> * The list is doubly linked (an extra (*prevconn) field is to be added into devif_callback_s structure).
> * The list has a head pointer (inside of outer net_driver_s structure), and a tail pointer is to be added into outer net_driver_s structure.
> * The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
> * The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL). If (*dev) argument is not NULL, the list tail can be taken from outer net_driver_s structure through (*dev) argument. Otherwise, the list tail can be found using a linear search (at least usrsock_setup_request_callback() and usrsock_pollsetup() always pass NULL pointer into (*dev) argument).
> * devif_callback_alloc() time complexity is O(1) or O(n) in the worst case (i.e. O(n) and O(n^2) to fill the whole list).
> * devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
> * devif_conn_event() time complexity is O(n).
> * Only three files are to be modified (netdev.h, devif.h and devif_callback.c). No need to adapt the whole network protocol stack.
> 
> **Approach 3.d:**
> 
> * The queue (list) is FIFO as it should be.
> * The list is doubly linked (an extra (*prevconn) field is to be added into devif_callback_s structure).
> * The list has a head pointer (inside of outer net_driver_s structure), and a tail pointer is to be added into outer net_driver_s structure.
> * The list item is devif_callback_s structure. It still has two pointers to two different list chains (*nxtconn and *nxtdev).
> * The first argument (*dev) of the list functions can be NULL, while the other argument (*list) is effective (not NULL).
> * An extra (*tail) argument is to be added to devif_callback_alloc() function.
> * devif_callback_alloc() time complexity is O(1) (i.e. O(n) to fill the whole list).
> * devif_callback_free() time complexity is O(1) (i.e. O(n) to empty the whole list).
> * devif_conn_event() time complexity is O(n).
> * Multiple files are to be modified. Though a risk to break something is low.
> 

It look like the best approach with the doubly linked list.

> Do you agree with all the details of the outlined explanation?

Yes, I agree.

Reply:
Why do you think Approach 4.s and Approach 3.d look the best?
Approach 3.s and Approach 3.d mostly correspond to each other.
Approach 4.s and Approach 4.d mostly correspond to each other.

Reply:
Sorry, my typo error. "Approach 4.s" should change to "Approach 3.s".

Reply:
Good. Approach 3.s and Approach 3.d look good to me as well.
First I can try Approach 3.s. If there are not some hidden issues, I can finally do Approach 3.d.

Reply:
The plan looks good, thanks you take additional time to improve the fix.

Reply:
Hi Xiang,
Please review my new commit of the intermediate step (Approach 3.s). Actually, I had to add (*tail) argument not only to devif_callback_alloc(), but to devif_conn_callback_free() as well.

Reply:
I'm sorry, just now I've added one missing line to icmp.h.